Init threadpool with user defined num_threads before default

janeyx99 opened the pull request:
Fixes #134714 (or attempts to, idk how to test yet)

For posterity, how one can test:
1. make sure you have USE_PTHREADPOOL=1 or pull a packaged binary
2. run gdb --args python, with `r` to enter, `Ctrl-C` to pause, and `c` to get back into Python
3. import torch
4. torch.set_num_threads(1), make sure this does not trigger any additional threads getting created.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136793

janeyx99 committed:
Init threadpool with user defined num_threads before default

[ghstack-poisoned]

albanD reviewed (approved):
Sounds good!
Please double check the behavior locally using gdb

janeyx99 committed:
Update on "Init threadpool with user defined num_threads before default"

Fixes #134714 (or attempts to, idk how to test yet)




[ghstack-poisoned]

albanD reviewed (approved):
Checked myself locally that this works

janeyx99 commented:
@pytorchbot merge

Thanks @albanD!

janeyx99 commented:
@pytorchbot revert -c ghfirst -m "Caused internal Oculus crash, and internal force landed a diff without exporting to GH =.="

janeyx99 committed:
Update on "Init threadpool with user defined num_threads before default"

Fixes #134714 (or attempts to, idk how to test yet)

For posterity, how one can test:
1. make sure you have USE_PTHREADPOOL=1 or pull a packaged binary
2. run gdb --args python, with `r` to enter, `Ctrl-C` to pause, and `c` to get back into Python
3. import torch
4. torch.set_num_threads(1), make sure this does not trigger any additional threads getting created.




[ghstack-poisoned]

janeyx99 commented:
Closing in favor of #137051
